cmd/service/deploy.go

```diff
@@ -24,11 +29,125 @@ To get more information, see the [deploy page from the documentation](https://do
 }
 
 func deployHandler(cmd *cobra.Command, args []string) {
-	service := prepareService(defaultPath(args))
-	reply, err := cli().DeployService(context.Background(), &core.DeployServiceRequest{
-		Service: service,
-	})
+	path := defaultPath(args)
+	serviceID, isValid, err := deployService(path)
+	if !isValid {
+		os.Exit(1)
+	}
 	utils.HandleError(err)
-	fmt.Println("Service deployed with ID:", aurora.Green(reply.ServiceID))
-	fmt.Println("To start it, run the command: mesg-core service start " + reply.ServiceID)
+
+	fmt.Println("Service deployed with ID:", aurora.Green(serviceID))
+	fmt.Printf("To start it, run the command: mesg-core service start %s\n", serviceID)
+
+}
+
+func deployService(path string) (serviceID string, isValid bool, err error) {
+	stream, err := cli().DeployService(context.Background())
+	if err != nil {
+		return "", true, err
+	}
+
+	deployment := make(chan deploymentResult)
+	go readDeployReply(stream, deployment)
+
+	if govalidator.IsURL(path) {
+		if err := stream.Send(&core.DeployServiceRequest{
+			Value: &core.DeployServiceRequest_Url{Url: path},
+		}); err != nil {
+			return "", true, err
+		}
+	} else {
+		if err := deployServiceSendServiceContext(path, stream); err != nil {
+			return "", true, err
+		}
+	}
+
+	if err := stream.CloseSend(); err != nil {
+		return "", true, err
+	}
+
+	result := <-deployment
+	return result.serviceID, result.isValid, result.err
+}
+
+func deployServiceSendServiceContext(path string, stream core.Core_DeployServiceClient) error {
+	archive, err := archive.TarWithOptions(path, &archive.TarOptions{
+		Compression: archive.Gzip,
+	})
+	if err != nil {
+		return err
+	}
+
+	buf := make([]byte, 1024)
```

Comment:
1ko seems a bit small. 

> Spoiler alert: it's 16KiB - 64KiB, according to oral tradition.
https://github.com/grpc/grpc.github.io/issues/371

Reply:
please check this out https://ops.tips/blog/sending-files-via-grpc/

Reply:
_It seems like at 1024bytes (1KB) we have a decent throughput._

_I was expecting something like 32KB: (1 << 15) to be the most optimal but it looks like it’s one of the worst scenarios regarding variance._